src/lib.rs

```diff
@@ -0,0 +1,83 @@
+use std::io;
+use tracing::{Level, Metadata};
+use tracing_subscriber::fmt::MakeWriter;
+
+enum BuildScriptWriterInner {
+    Informational(io::Stderr),
+    ErrorsAndWarnings { first_write: bool, writer: io::Stdout },
+}
+
+/// A writer intended to support the [output capturing of build scripts](https://doc.rust-lang.org/cargo/reference/build-scripts.html#outputs-of-the-build-script).
+/// `BuildScriptWriter` can be used by [`tracing_subscriber::fmt::Subscriber`](tracing_subscriber::fmt::Subscriber) or [`tracing_subscriber::fmt::Layer`](tracing_subscriber::fmt::Layer)
+/// to enable capturing output in build scripts.
+pub struct BuildScriptWriter(BuildScriptWriterInner);
+
+impl BuildScriptWriter {
+    /// Create a writer for informational events.
+    /// Events will be written to stderr.
+    pub fn informational() -> Self {
+        Self(BuildScriptWriterInner::Informational(io::stderr()))
+    }
+
+    /// Create a writer for warning and error events.
+    /// Events will be written to stdout after having `cargo::warning=` prepended.
+    pub fn errors_and_warnings() -> Self {
+        Self(BuildScriptWriterInner::ErrorsAndWarnings { first_write: true, writer: io::stdout() })
+    }
+}
+
+impl io::Write for BuildScriptWriter {
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        match &mut self.0 {
+            BuildScriptWriterInner::Informational(writer) => writer.write(buf),
+            BuildScriptWriterInner::ErrorsAndWarnings { first_write, writer } => {
+                if *first_write {
+                    writer.write(b"cargo::warning=")?;
+                    *first_write = false;
+                }
```

Comment:
I might be missing something, but shouldn't every line start with this?

I think all lines inside the event should be escaped and every event should be prepended with the warning directive, no?

Reply:
No, you are completely, right. I wasn't thinking of multi-line log messages 😅

Reply:
Should be resolved now I think.

I've also found a way to add some test cases via a separate crate with build script